### PR TITLE
feat: metadata protocol

### PR DESCRIPTION
**Pull request by danisharora099**
This PR introduces the Metadata Protocol as a service

## Notes: 
- [x] dependent on https://github.com/waku-org/js-waku/pull/1740
- resolves https://github.com/waku-org/js-waku/issues/1652

### CODE REVIEW

File: packages/core/src/lib/metadata/index.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    );
+
+    const bytes = new Uint8ArrayList();
+    res.forEach((chunk) => {
+      bytes.append(chunk);
+    });
+
+    const shardInfoRes = proto_metadata.WakuMetadataResponse.decode(bytes);
+    if (!shardInfoRes) {
+      throw new Error("WakuMetadata response is undefined");
```
[the remark this answers is not in the document]
Comment (danisharora099) [reply packages/core/src/lib/metadata/index.ts:50]: it was consciously done as I'd prefer that the mount error leads to node creation error, than the node not working as expected.
addressed your suggestion now, and perhaps something we can change in the future if required

File: packages/core/src/lib/metadata/index.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      lp.decode,
+      async (source) => await all(source)
+    );
+
+    const bytes = new Uint8ArrayList();
+    res.forEach((chunk) => {
+      bytes.append(chunk);
+    });
+
+    const shardInfoRes = proto_metadata.WakuMetadataResponse.decode(bytes);
```
[the remark this answers is not in the document]
Comment (danisharora099) [reply packages/core/src/lib/metadata/index.ts:48]: good spot. addressed: deef4973e6d183d12bb9ae53f3fd49b0b9761e42

File: packages/core/src/lib/metadata/index.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  private async onRequest(streamData: IncomingStreamData): Promise<void> {
+    const encodedRpcQuery = proto_metadata.WakuMetadataRequest.encode(
+      this.shardInfo
+    );
+
+    const res = await pipe(
+      [encodedRpcQuery],
+      lp.encode,
+      streamData.stream,
+      lp.decode,
```
[the remark this answers is not in the document]
Comment (danisharora099) [reply packages/core/src/lib/metadata/index.ts:39]: addressed: https://github.com/waku-org/js-waku/pull/1732#discussion_r1410659192

File: packages/tests/tests/metadata.spec.ts [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+    const nwaku1Ma = await nwaku1.getMultiaddrWithId();
+    const nwaku1PeerId = await nwaku1.getPeerId();
+
+    waku = await createLightNode({ shardInfo: shardInfo2 });
+    await waku.start();
+    await waku.libp2p.dialProtocol(nwaku1Ma, MetadataCodec);
+
+    await expect(
+      waku.libp2p.services.metadata?.query(nwaku1PeerId)
```
[the remark this answers is not in the document]
Comment (danisharora099) [reply packages/tests/tests/metadata.spec.ts:127]: the metadata query fails (with a throw) if the clusters are different between the service node it's attempting to dial therefore the expecting a catch

you're right in pointing out that there will no active connection, added another expect that also checks for that in all test cases: d53eddbda95c595df83049a207dbf1ad7845f3b3

File: packages/core/src/lib/metadata/index.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  /**
+   * Handle an incoming metadata request
+   */
+  private async onRequest(streamData: IncomingStreamData): Promise<void> {
+    try {
+      const encodedResponse = proto_metadata.WakuMetadataResponse.encode(
+        this.shardInfo
+      );
+
+      await pipe([encodedResponse], lp.encode, streamData.stream);
```
[the remark this answers is not in the document]
Comment (danisharora099) [reply packages/core/src/lib/metadata/index.ts]: yes, that part was intended be tracked part of https://github.com/waku-org/js-waku/issues/1684 but addressed it here and actually makes more sense! thanks:) d0c9683c64e4af4a0daaf082f23ac457b6af878f

File: packages/tests/tests/metadata.spec.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (fryorcraken) [reply packages/tests/tests/metadata.spec.ts]: LGTM

**Pull request merged**